<xhash>: dllexport emits error C2338: This function only works with multi containers

```
C:\Temp>type meow.cpp
```
```cpp
#include <unordered_set>
using namespace std;

struct __declspec(dllexport) SymTbl : unordered_set<int> {};
```
```
C:\Temp>cl /EHsc /nologo /W4 /MD /std:c++17 /c /wd4251 meow.cpp
meow.cpp
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.28.29828\include\xhash(1996): error C2338: This function only works with multi containers
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.28.29828\include\xhash(1995): note: while compiling class template member function 'bool std::_Hash<std::_Uset_traits<_Kty,std::_Uhash_compare<_Kty,_Hasher,_Keyeq>,_Alloc,false>>::_Multi_equal(const std::_Hash<std::_Uset_traits<_Kty,std::_Uhash_compare<_Kty,_Hasher,_Keyeq>,_Alloc,false>> &) const'
        with
        [
            _Kty=int,
            _Hasher=std::hash<int>,
            _Keyeq=std::equal_to<int>,
            _Alloc=std::allocator<int>
        ]
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.28.29828\include\unordered_set(65): note: see reference to class template instantiation 'std::_Hash<std::_Uset_traits<_Kty,std::_Uhash_compare<_Kty,_Hasher,_Keyeq>,_Alloc,false>>' being compiled
        with
        [
            _Kty=int,
            _Hasher=std::hash<int>,
            _Keyeq=std::equal_to<int>,
            _Alloc=std::allocator<int>
        ]
meow.cpp(4): note: see reference to class template instantiation 'std::unordered_set<int,std::hash<int>,std::equal_to<int>,std::allocator<int>>' being compiled
```

This is because we have:
https://github.com/microsoft/STL/blob/47cd70315d560e4fe1ac699c0eb8ec8bfc5cb658/stl/inc/xhash#L1844-L1845
called by only:
https://github.com/microsoft/STL/blob/47cd70315d560e4fe1ac699c0eb8ec8bfc5cb658/stl/inc/xhash#L1949-L1950

The Standard says that `_Multi_equal()` will be instantiated "on demand" only when it's called, but dllexport tries to instantiate all non-template member functions.

Exporting classes that derive from STL containers isn't exactly recommended, but we should avoid this code pattern that causes compiler errors.

Reported as DevCom-1319981 and Microsoft-internal VSO-1270072 / AB#1270072 .

--- a/stl/inc/xhash
+++ b/stl/inc/xhash
@@ -1841,6 +1841,7 @@ protected:
         }
     }
 
+    template <bool _Multi2 = _Traits::_Multi, enable_if_t<_Multi2, int> = 0>
     _NODISCARD bool _Multi_equal(const _Hash& _Right) const {
         static_assert(_Traits::_Multi, "This function only works with multi containers");
         _STL_INTERNAL_CHECK(this->size() == _Right.size());

--- a/tests/std/test.lst
+++ b/tests/std/test.lst
@@ -173,6 +173,7 @@ tests\GH_001105_custom_streambuf_throws
 tests\GH_001123_random_cast_out_of_range
 tests\GH_001411_core_headers
 tests\GH_001541_case_sensitive_boolalpha
+tests\GH_001638_dllexport_derived_classes
 tests\LWG2597_complex_branch_cut
 tests\LWG3018_shared_ptr_function
 tests\P0019R8_atomic_ref

--- a/tests/std/tests/GH_001638_dllexport_derived_classes/env.lst
+++ b/tests/std/tests/GH_001638_dllexport_derived_classes/env.lst
@@ -1,0 +1,4 @@
+# Copyright (c) Microsoft Corporation.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+RUNALL_INCLUDE ..\usual_matrix.lst

--- a/tests/std/tests/GH_001638_dllexport_derived_classes/env.lst
+++ b/tests/std/tests/GH_001638_dllexport_derived_classes/env.lst
@@ -1,4 +1,4 @@
 # Copyright (c) Microsoft Corporation.
 # SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 
-RUNALL_INCLUDE ..\usual_matrix.lst
+RUNALL_INCLUDE ..\impure_matrix.lst

--- a/tests/std/tests/GH_001638_dllexport_derived_classes/test.compile.pass.cpp
+++ b/tests/std/tests/GH_001638_dllexport_derived_classes/test.compile.pass.cpp
@@ -1,0 +1,52 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <array>
+#include <deque>
+#include <forward_list>
+#include <list>
+#include <map>
+#include <queue>
+#include <set>
+#include <stack>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#if _HAS_CXX20
+#include <span>
+#endif // _HAS_CXX20
+
+using namespace std;
+
+int main() {} // COMPILE-ONLY
+
+#pragma warning(disable : 4251) // class 'A' needs to have dll-interface to be used by clients of class 'B'
+#pragma warning(disable : 4275) // non dll-interface struct 'A' used as base for dll-interface class 'B'
+
+struct __declspec(dllexport) ExportedArray : array<int, 3> {};
+struct __declspec(dllexport) ExportedArrayZero : array<int, 0> {};
+struct __declspec(dllexport) ExportedDeque : deque<int> {};
+struct __declspec(dllexport) ExportedForwardList : forward_list<int> {};
+struct __declspec(dllexport) ExportedList : list<int> {};
+struct __declspec(dllexport) ExportedVector : vector<int> {};
+struct __declspec(dllexport) ExportedVectorBool : vector<bool> {};
+
+struct __declspec(dllexport) ExportedMap : map<int, int> {};
+struct __declspec(dllexport) ExportedMultimap : multimap<int, int> {};
+struct __declspec(dllexport) ExportedSet : set<int> {};
+struct __declspec(dllexport) ExportedMultiset : multiset<int> {};
+
+struct __declspec(dllexport) ExportedUnorderedMap : unordered_map<int, int> {};
+struct __declspec(dllexport) ExportedUnorderedMultimap : unordered_multimap<int, int> {};
+struct __declspec(dllexport) ExportedUnorderedSet : unordered_set<int> {};
+struct __declspec(dllexport) ExportedUnorderedMultiset : unordered_multiset<int> {};
+
+struct __declspec(dllexport) ExportedQueue : queue<int> {};
+struct __declspec(dllexport) ExportedPriorityQueue : priority_queue<int> {};
+struct __declspec(dllexport) ExportedStack : stack<int> {};
+
+#if _HAS_CXX20
+struct __declspec(dllexport) ExportedSpan : span<int> {};
+struct __declspec(dllexport) ExportedSpanThree : span<int, 3> {};
+#endif // _HAS_CXX20

